pkg/server/commands/reverseexpand/reverse_expand_test.go

```diff
@@ -443,3 +447,277 @@ func TestReverseExpandIgnoresInvalidTuples(t *testing.T) {
+				Throttler:    mockThrottler,
+				Threshold:    200,
+				MaxThreshold: 0,
+			}),
+		)
+		mockThrottler.EXPECT().Throttle(gomock.Any()).Times(0)
+		dispatchCountValue := uint32(190)
+		metadata := NewResolutionMetadata()
+		metadata.DispatchCounter.Store(dispatchCountValue)
+
+		reverseExpandQuery.throttle(ctx, metadata)
+	})
+
+	t.Run("dispatch_should_use_request_threshold_if_available", func(t *testing.T) {
+		mockThrottler := mocks.NewMockThrottler(mockController)
+		reverseExpandQuery := NewReverseExpandQuery(
+			mockDatastore,
+			ts,
+			WithDispatchThrottlerConfig(threshold.Config{
+				Throttler:    mockThrottler,
+				Threshold:    0,
+				MaxThreshold: 210,
+			}),
+		)
+		mockThrottler.EXPECT().Throttle(gomock.Any()).Times(1)
+		dispatchCountValue := uint32(201)
+		ctx := context.Background()
+		ctx = threshold.ContextWithDispatchThrottlingThreshold(ctx, 200)
+		metadata := NewResolutionMetadata()
+		metadata.DispatchCounter.Store(dispatchCountValue)
+
+		reverseExpandQuery.throttle(ctx, metadata)
+	})
+
+	t.Run("should_respect_max_threshold", func(t *testing.T) {
+		mockThrottler := mocks.NewMockThrottler(mockController)
+		reverseExpandQuery := NewReverseExpandQuery(
+			mockDatastore,
+			ts,
+			WithDispatchThrottlerConfig(threshold.Config{
+				Throttler:    mockThrottler,
+				Threshold:    200,
+				MaxThreshold: 300,
+			}),
+		)
+		mockThrottler.EXPECT().Throttle(gomock.Any()).Times(1)
+		dispatchCountValue := uint32(301)
+		ctx := context.Background()
+		ctx = threshold.ContextWithDispatchThrottlingThreshold(ctx, 1000)
+		metadata := NewResolutionMetadata()
+		metadata.DispatchCounter.Store(dispatchCountValue)
+
+		reverseExpandQuery.throttle(ctx, metadata)
+	})
+}
+
+func TestReverseExpandDispatchCount(t *testing.T) {
+	ds := memory.New()
+	t.Cleanup(ds.Close)
+	ctx := storage.ContextWithRelationshipTupleReader(context.Background(), ds)
+	ctrl := gomock.NewController(t)
+	mockThrottler := mocks.NewMockThrottler(ctrl)
+	t.Cleanup(ctrl.Finish)
+	tests := []struct {
+		name                    string
+		model                   string
+		tuples                  []string
+		objectType              string
+		relation                string
+		user                    *UserRefObject
+		throttlingEnabled       bool
+		expectedDispatchCount   uint32
+		expectedThrottlingValue int
+		expectedWasThrottled    bool
+	}{
+		{
+			name: "should_throttle",
+			model: `model
+			schema 1.1
+
+			type user
+
+			type folder
+				 relations
+					  define editor: [user]
+					  define viewer: [user] or editor 
+			`,
+			tuples: []string{
+				"folder:C#editor@user:jon",
+				"folder:B#viewer@user:jon",
+				"folder:A#viewer@user:jon",
+			},
+			objectType:              "folder",
+			relation:                "viewer",
+			user:                    &UserRefObject{Object: &openfgav1.Object{Type: "user", Id: "jon"}},
+			throttlingEnabled:       true,
+			expectedWasThrottled:    true,
+			expectedDispatchCount:   4,
```

Comment:
This isn't directly related to this code here in this PR, but we need to follow up on this. This value isn't correct.

In this example there aren't any dispatches. A dispatch isn't a direct lookup - it's an evaluation of another object relationship with respect to the incoming one. In this case we have 2 direct reverse lookups - one for the edge for `folder#viewer` and one for the edge for `folder#editor` for the subject `user:jon`. This will result in 2 database lookups, but no usersets are involved whatsoever and therefore no dispatches should occur. This should refer to datastore query count and its value should be 2.

Reply:
We will follow up.